src/nfa/operations.cc

```diff
@@ -114,6 +114,364 @@ namespace {
+        covering_states.push_back(mata::utils::OrdVector<State>());
+        covering_indexes.push_back(mata::utils::OrdVector<State>());
+
+        while (it != subset_map->end()) {               // goes through all found states
+            if (it->first.IsSubsetOf(T)) {
+                // check if T is covered
+                // if so add covering state to its covering StateSet
+
+                covering_states[Tid].insert(it->first);
+                covering_indexes[Tid].insert(it->second);
+            }
+            else if (T.IsSubsetOf(it->first)) {
+                // check if state in map is covered
+                // if so add covering state to its covering StateSet
+
+                covering_states[it->second].insert(T);
+                covering_indexes[it->second].insert(Tid);
+
+                // check is some already existing state that had a new covering state added turned fully covered
+                if (it->first == covering_states[it->second]) {
+                    // if any covered state is in the covering set of newly turned covered state,
+                    // then it has to be replaced by its covering set
+                    //
+                    // same applies for any covered state, if it contains newly turned state in theirs
+                    // covering set, then it has to be updated
+                    for (auto covered_pair: covered) {
+                        if (covering_indexes[covered_pair.second].contains(it->second)) {
+                            covering_indexes[covered_pair.second].erase(it->second);
+                            covering_indexes[covered_pair.second].insert(covering_indexes[it->second]);
+                        }
+                        if (covering_indexes[it->second].contains(covered_pair.second)) {
+                            covering_indexes[it->second].erase(covered_pair.second);
+                            covering_indexes[it->second].insert(covering_indexes[covered_pair.second]);
+                        }
+                    }
+
+                    // remove covered state from the automaton, replace with covering set
+                    remove_covered_state(covering_indexes[it->second], it->second, result);
+
+                    std::unordered_map<StateSet, State>::iterator temp = it++;
+                    // move state from subset_map to covered
+                    auto transfer = subset_map->extract(temp);
+                    covered.insert(std::move(transfer));
+                    continue;           // skip increasing map pointer
+                }
+            }
+            ++it;
+        }
+    }
+
+    Nfa residual_with(const Nfa& aut) {         // modified algorithm of determinization
+
+        Nfa result;
+
+        //assuming all sets targets are non-empty
+        std::vector<std::pair<State, StateSet>> worklist;
+        std::unordered_map<StateSet, State> * subset_map = new std::unordered_map<StateSet,State>();
```

Comment:
On the other hand I could add back the possibility to get the `subset_map` back from the function as it is for determinization, but I don't expect that it would be utilized, as the function most likely won't be called directly by the user. The automaton is also trimmed at the end of the reduction so the `subset_map` does not have to contain the correct numbering of the states.

Reply:
If there is a memory leak that you have found out, great. It would help us immensely if you could open an issue, ideally with a reproducible example.

If the dynamic memory allocation could be removed from here, it would be better in my opinion. In this case, the dynamic memory structure is a local variable entirely enclosed in the function. I see no need to create the structure dynamically. I would suggest rewriting this code without dynamic memory. Case in point, the memory leaks mentioned above :)

Since the `subset_map` cannot give any useful information, I would not bother adding in into the interface as a parameter. The interface is bound to change sometime in the future anyway.

Reply:
Sure, will open an issue about that and remove the dynamic memory allocation as it is not required.